Unable to navigate through the job execution pages

Also you can't navigate through the job execution pages anymore. 

org.springframework.jdbc.BadSqlGrammarException: PreparedStatementCallback; bad SQL grammar [SELECT E.JOB_EXECUTION_ID FROM (SELECT E.JOB_EXECUTION_ID, ROWNUM as TMP_ROW_NUM FROM (SELECT E.JOB_EXECUTION_ID FROM AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET JOIN AGGREGATE_TASK_BATCH B ON E.JOB_EXECUTION_ID = B.JOB_EXECUTION_ID AND E.SCHEMA_TARGET = B.SCHEMA_TARGET JOIN AGGREGATE_TASK_EXECUTION T ON B.TASK_EXECUTION_ID = T.TASK_EXECUTION_ID AND B.SCHEMA_TARGET = T.SCHEMA_TARGET ORDER BY E.JOB_EXECUTION_ID DESC)) WHERE TMP_ROW_NUM = 20]; nested exception is java.sql.SQLSyntaxErrorException: ORA-00904: \"E\".\"JOB_EXECUTION_ID\": ungültige ID

_Originally posted by @klopfdreh in https://github.com/spring-cloud/spring-cloud-dataflow/issues/5524#issuecomment-1774833293_

Full stacktrace: 
```
org.springframework.jdbc.BadSqlGrammarException: PreparedStatementCallback; bad SQL grammar [SELECT E.JOB_EXECUTION_ID FROM (SELECT E.JOB_EXECUTION_ID, ROWNUM as TMP_ROW_NUM FROM (SELECT E.JOB_EXECUTION_ID FROM AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET JOIN AGGREGATE_TASK_BATCH B ON E.JOB_EXECUTION_ID = B.JOB_EXECUTION_ID AND E.SCHEMA_TARGET = B.SCHEMA_TARGET JOIN AGGREGATE_TASK_EXECUTION T ON B.TASK_EXECUTION_ID = T.TASK_EXECUTION_ID AND B.SCHEMA_TARGET = T.SCHEMA_TARGET ORDER BY E.JOB_EXECUTION_ID DESC)) WHERE TMP_ROW_NUM = 20]; nested exception is java.sql.SQLSyntaxErrorException: ORA-00904: "E"."JOB_EXECUTION_ID": invalid identifier

	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.doTranslate(SQLErrorCodeSQLExceptionTranslator.java:236)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:73)
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1577)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:669)
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:715)
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:746)
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:759)
	at org.springframework.jdbc.core.JdbcTemplate.queryForObject(JdbcTemplate.java:881)
	at org.springframework.jdbc.core.JdbcTemplate.queryForObject(JdbcTemplate.java:908)
	at org.springframework.cloud.dataflow.server.repository.JdbcAggregateJobQueryDao.queryForProvider(JdbcAggregateJobQueryDao.java:607)
	at org.springframework.cloud.dataflow.server.repository.JdbcAggregateJobQueryDao.getJobExecutionsWithStepCount(JdbcAggregateJobQueryDao.java:516)
	at org.springframework.cloud.dataflow.server.repository.JdbcAggregateJobQueryDao.listJobExecutionsWithStepCount(JdbcAggregateJobQueryDao.java:336)
	at org.springframework.cloud.dataflow.server.service.impl.DefaultTaskJobService.listJobExecutionsWithStepCount(DefaultTaskJobService.java:132)
	at org.springframework.cloud.dataflow.server.service.impl.DefaultTaskJobService$$FastClassBySpringCGLIB$$d486827e.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:792)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:762)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:762)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:707)
	at org.springframework.cloud.dataflow.server.service.impl.DefaultTaskJobService$$EnhancerBySpringCGLIB$$7f975b34.listJobExecutionsWithStepCount(<generated>)
	at org.springframework.cloud.dataflow.server.controller.JobExecutionThinController.listJobsOnly(JobExecutionThinController.java:98)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1072)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:965)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:529)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:623)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:209)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:96)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:168)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:481)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:390)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:928)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1794)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.sql.SQLSyntaxErrorException: ORA-00904: "E"."JOB_EXECUTION_ID": invalid identifier

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:629)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:563)
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1150)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:770)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:298)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:497)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:151)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:936)
	at oracle.jdbc.driver.OracleStatement.prepareDefineBufferAndExecute(OracleStatement.java:1171)
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:1100)
	at oracle.jdbc.driver.OracleStatement.executeSQLSelect(OracleStatement.java:1425)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1308)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3745)
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:3854)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1097)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java)
	at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:724)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:650)
	... 79 common frames omitted
Caused by: oracle.jdbc.OracleDatabaseException: ORA-00904: "E"."JOB_EXECUTION_ID": invalid identifier

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:636)
	... 97 common frames omitted
```

Test run appears that this is an issue w/ not only Oracle, but also for...

**DB2**
```
DB2SmokeTest>AbstractSmokeTest.shouldListJobExecutionsUsingPerformantRowNumberQuery:171 » BadSqlGrammar PreparedStatementCallback; bad SQL grammar [SELECT TMP_SUB.JOB_EXECUTION_ID FROM ( SELECT E.JOB_EXECUTION_ID, ROW_NUMBER() OVER ( ORDER BY JOB_EXECUTION_ID DESC) AS ROW_NUMBER FROM AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET JOIN AGGREGATE_TASK_BATCH B ON E.JOB_EXECUTION_ID = B.JOB_EXECUTION_ID AND E.SCHEMA_TARGET = B.SCHEMA_TARGET JOIN AGGREGATE_TASK_EXECUTION T ON B.TASK_EXECUTION_ID = T.TASK_EXECUTION_ID AND B.SCHEMA_TARGET = T.SCHEMA_TARGET) AS TMP_SUB WHERE TMP_SUB.ROW_NUMBER = 2 ORDER BY TMP_SUB.JOB_EXECUTION_ID DESC]; nested exception is com.ibm.db2.jcc.am.SqlSyntaxErrorException: DB2 SQL Error: SQLCODE=-203, SQLSTATE=42702, SQLERRMC=JOB_EXECUTION_ID, DRIVER=4.32.28
2023-12-15T18:11:03.5395334Z [ERROR]   
```

**SQLServer2017**
```
SqlServer2017SmokeTest>AbstractSmokeTest.shouldListJobExecutionsUsingPerformantRowNumberQuery:171 » BadSqlGrammar PreparedStatementCallback; bad SQL grammar [SELECT TMP_SUB.JOB_EXECUTION_ID FROM ( SELECT E.JOB_EXECUTION_ID, ROW_NUMBER() OVER ( ORDER BY JOB_EXECUTION_ID DESC) AS ROW_NUMBER FROM AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET JOIN AGGREGATE_TASK_BATCH B ON E.JOB_EXECUTION_ID = B.JOB_EXECUTION_ID AND E.SCHEMA_TARGET = B.SCHEMA_TARGET JOIN AGGREGATE_TASK_EXECUTION T ON B.TASK_EXECUTION_ID = T.TASK_EXECUTION_ID AND B.SCHEMA_TARGET = T.SCHEMA_TARGET) AS TMP_SUB WHERE TMP_SUB.ROW_NUMBER = 2 ORDER BY TMP_SUB.JOB_EXECUTION_ID DESC]; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Ambiguous column name 'JOB_EXECUTION_ID'.
```

**SQLServer2019**
```
SqlServer2019SmokeTest>AbstractSmokeTest.shouldListJobExecutionsUsingPerformantRowNumberQuery:171 » BadSqlGrammar PreparedStatementCallback; bad SQL grammar [SELECT TMP_SUB.JOB_EXECUTION_ID FROM ( SELECT E.JOB_EXECUTION_ID, ROW_NUMBER() OVER ( ORDER BY JOB_EXECUTION_ID DESC) AS ROW_NUMBER FROM AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET JOIN AGGREGATE_TASK_BATCH B ON E.JOB_EXECUTION_ID = B.JOB_EXECUTION_ID AND E.SCHEMA_TARGET = B.SCHEMA_TARGET JOIN AGGREGATE_TASK_EXECUTION T ON B.TASK_EXECUTION_ID = T.TASK_EXECUTION_ID AND B.SCHEMA_TARGET = T.SCHEMA_TARGET) AS TMP_SUB WHERE TMP_SUB.ROW_NUMBER = 2 ORDER BY TMP_SUB.JOB_EXECUTION_ID DESC]; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Ambiguous column name 'JOB_EXECUTION_ID'.
```
**SQLServer2022**
```
SqlServer2022SmokeTest>AbstractSmokeTest.shouldListJobExecutionsUsingPerformantRowNumberQuery:171 » BadSqlGrammar PreparedStatementCallback; bad SQL grammar [SELECT TMP_SUB.JOB_EXECUTION_ID FROM ( SELECT E.JOB_EXECUTION_ID, ROW_NUMBER() OVER ( ORDER BY JOB_EXECUTION_ID DESC) AS ROW_NUMBER FROM AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET JOIN AGGREGATE_TASK_BATCH B ON E.JOB_EXECUTION_ID = B.JOB_EXECUTION_ID AND E.SCHEMA_TARGET = B.SCHEMA_TARGET JOIN AGGREGATE_TASK_EXECUTION T ON B.TASK_EXECUTION_ID = T.TASK_EXECUTION_ID AND B.SCHEMA_TARGET = T.SCHEMA_TARGET) AS TMP_SUB WHERE TMP_SUB.ROW_NUMBER = 2 ORDER BY TMP_SUB.JOB_EXECUTION_ID DESC]; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Ambiguous column name 'JOB_EXECUTION_ID'.
```

--- a/README.md
+++ b/README.md
@@ -113,6 +113,19 @@ You can follow the steps in the [MSSQL on Mac ARM64](https://github.com/spring-c
 
 ----
 
+## Running Locally w/ IBM DB2
+By default, the Dataflow server jar does not include the DB2 database driver dependency.
+If you want to use DB2 for development/testing when running locally, you can specify the `local-dev-db2` Maven profile when building.
+The following command will include the DB2 driver dependency in the jar:
+```
+$ ./mvnw -s .settings.xml clean package -Plocal-dev-db2
+```
+You can follow the steps in the [DB2 on Mac ARM64](https://github.com/spring-cloud/spring-cloud-dataflow/wiki/DB2-on-Mac-ARM64#running-dataflow-locally-against-db2) Wiki to run DB2 locally in Docker with Dataflow pointing at it.
+
+> **NOTE:** If you are not running Mac ARM64 just skip the steps related to Homebrew and Colima
+
+----
+
 ## Contributing
 
 We welcome contributions! See the [CONTRIBUTING](./CONTRIBUTING.adoc) guide for details.

--- a/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java
+++ b/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java
@@ -49,6 +49,7 @@ import org.springframework.batch.core.repository.dao.JdbcJobExecutionDao;
 import org.springframework.batch.item.database.Order;
 import org.springframework.batch.item.database.PagingQueryProvider;
 import org.springframework.batch.item.database.support.AbstractSqlPagingQueryProvider;
+import org.springframework.batch.item.database.support.Db2PagingQueryProvider;
 import org.springframework.batch.item.database.support.OraclePagingQueryProvider;
 import org.springframework.batch.item.database.support.SqlPagingQueryProviderFactoryBean;
 import org.springframework.batch.item.database.support.SqlPagingQueryUtils;
@@ -841,8 +842,8 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 	}
 
 	/**
-	 * A {@link SqlPagingQueryProviderFactoryBean} specialization that overrides the {@code Oracle} paging
-	 * {@link SafeOraclePagingQueryProvider provider} with an implementation that properly handles sort aliases.
+	 * A {@link SqlPagingQueryProviderFactoryBean} specialization that overrides the {@code Oracle, MSSQL, and DB2}
+	 * paging {@link SafeOraclePagingQueryProvider provider} with an implementation that properly handles sort aliases.
 	 * <p><b>NOTE:</b> nested within the aggregate DAO as this is the only place that needs this specialization.
 	 */
 	static class SafeSqlPagingQueryProviderFactoryBean extends SqlPagingQueryProviderFactoryBean {
@@ -863,6 +864,9 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 			}
 			else if (provider instanceof SqlServerPagingQueryProvider) {
 				provider = new SafeSqlServerPagingQueryProvider((SqlServerPagingQueryProvider) provider, this.dataSource);
+			}
+			else if (provider instanceof Db2PagingQueryProvider) {
+				provider = new SafeDb2PagingQueryProvider((Db2PagingQueryProvider) provider, this.dataSource);
 			}
 			return provider;
 		}
@@ -967,6 +971,43 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 	static class SafeSqlServerPagingQueryProvider extends SqlServerPagingQueryProvider {
 
 		SafeSqlServerPagingQueryProvider(SqlServerPagingQueryProvider delegate, DataSource dataSource) {
+			// Have to use reflection to retrieve the provider fields
+			this.setFromClause(extractField(delegate, "fromClause", String.class));
+			this.setWhereClause(extractField(delegate, "whereClause", String.class));
+			this.setSortKeys(extractField(delegate, "sortKeys", Map.class));
+			this.setSelectClause(extractField(delegate, "selectClause", String.class));
+			this.setGroupClause(extractField(delegate, "groupClause", String.class));
+			try {
+				this.init(dataSource);
+			}
+			catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		private <T> T extractField(AbstractSqlPagingQueryProvider target, String fieldName, Class<T> fieldType) {
+			Field field = ReflectionUtils.findField(AbstractSqlPagingQueryProvider.class, fieldName, fieldType);
+			ReflectionUtils.makeAccessible(field);
+			return (T) ReflectionUtils.getField(field, target);
+		}
+
+		@Override
+		protected String getOverClause() {
+			// Overrides the parent impl to use 'getSortKeys' instead of 'getSortKeysWithoutAliases'
+			StringBuilder sql = new StringBuilder();
+			sql.append(" ORDER BY ").append(SqlPagingQueryUtils.buildSortClause(this.getSortKeys()));
+			return sql.toString();
+		}
+
+	}
+
+	/**
+	 * A {@link Db2PagingQueryProvider paging provider} for {@code DB2} that works around the fact that the
+	 * DB2 provider in Spring Batch 4.x does not properly handle sort aliases when generating jump to page queries.
+	 */
+	static class SafeDb2PagingQueryProvider extends Db2PagingQueryProvider {
+
+		SafeDb2PagingQueryProvider(Db2PagingQueryProvider delegate, DataSource dataSource) {
 			// Have to use reflection to retrieve the provider fields
 			this.setFromClause(extractField(delegate, "fromClause", String.class));
 			this.setWhereClause(extractField(delegate, "whereClause", String.class));

--- a/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java
+++ b/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java
@@ -818,8 +818,7 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 		}
 		if (fields.contains("E.JOB_EXECUTION_ID") && this.useRowNumberOptimization) {
 			Order order = sortKeys.get("E.JOB_EXECUTION_ID");
-			String orderString = Optional.ofNullable(order)
-					.map(orderKey -> orderKey == Order.DESCENDING ? "DESC" : "ASC").orElse("DESC");
+			String orderString = (order == null || order == Order.DESCENDING) ? "DESC" : "ASC";
 			fields += ", ROW_NUMBER() OVER (PARTITION BY E.JOB_EXECUTION_ID ORDER BY E.JOB_EXECUTION_ID " + orderString + ") as RN";
 		}
 		factory.setSelectClause(fields);

--- a/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java
+++ b/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcAggregateJobQueryDao.java
@@ -16,6 +16,7 @@
 
 package org.springframework.cloud.dataflow.server.repository;
 
+import java.lang.reflect.Field;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.time.Instant;
@@ -47,7 +48,11 @@ import org.springframework.batch.core.launch.NoSuchJobInstanceException;
 import org.springframework.batch.core.repository.dao.JdbcJobExecutionDao;
 import org.springframework.batch.item.database.Order;
 import org.springframework.batch.item.database.PagingQueryProvider;
+import org.springframework.batch.item.database.support.AbstractSqlPagingQueryProvider;
+import org.springframework.batch.item.database.support.OraclePagingQueryProvider;
 import org.springframework.batch.item.database.support.SqlPagingQueryProviderFactoryBean;
+import org.springframework.batch.item.database.support.SqlPagingQueryUtils;
+import org.springframework.batch.item.database.support.SqlServerPagingQueryProvider;
 import org.springframework.cloud.dataflow.core.DataFlowPropertyKeys;
 import org.springframework.cloud.dataflow.core.database.support.DatabaseType;
 import org.springframework.cloud.dataflow.rest.job.JobInstanceExecutions;
@@ -75,6 +80,7 @@ import org.springframework.jdbc.core.RowCallbackHandler;
 import org.springframework.jdbc.core.RowMapper;
 import org.springframework.util.Assert;
 import org.springframework.util.ObjectUtils;
+import org.springframework.util.ReflectionUtils;
 import org.springframework.util.StringUtils;
 
 /**
@@ -802,7 +808,7 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 	 * @throws Exception if page provider is not created.
 	 */
 	private PagingQueryProvider getPagingQueryProvider(String fields, String fromClause, String whereClause, Map<String, Order> sortKeys) throws Exception {
-		SqlPagingQueryProviderFactoryBean factory = new SqlPagingQueryProviderFactoryBean();
+		SqlPagingQueryProviderFactoryBean factory = new SafeSqlPagingQueryProviderFactoryBean();
 		factory.setDataSource(dataSource);
 		fromClause = "AGGREGATE_JOB_INSTANCE I JOIN AGGREGATE_JOB_EXECUTION E ON I.JOB_INSTANCE_ID=E.JOB_INSTANCE_ID AND I.SCHEMA_TARGET=E.SCHEMA_TARGET" + (fromClause == null ? "" : " " + fromClause);
 		factory.setFromClause(fromClause);
@@ -811,7 +817,8 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 		}
 		if (fields.contains("E.JOB_EXECUTION_ID") && this.useRowNumberOptimization) {
 			Order order = sortKeys.get("E.JOB_EXECUTION_ID");
-			String orderString = Optional.ofNullable(order).map(orderKey -> orderKey == Order.DESCENDING ? "DESC" : "ASC").orElse("DESC");
+			String orderString = Optional.ofNullable(order)
+					.map(orderKey -> orderKey == Order.DESCENDING ? "DESC" : "ASC").orElse("DESC");
 			fields += ", ROW_NUMBER() OVER (PARTITION BY E.JOB_EXECUTION_ID ORDER BY E.JOB_EXECUTION_ID " + orderString + ") as RN";
 		}
 		factory.setSelectClause(fields);
@@ -831,5 +838,162 @@ public class JdbcAggregateJobQueryDao implements AggregateJobQueryDao {
 			LOG.warn("Unable to determine if DB supports ROW_NUMBER() function (reason: " + e.getMessage() + ")", e);
 		}
 		return false;
+	}
+
+	/**
+	 * A {@link SqlPagingQueryProviderFactoryBean} specialization that overrides the {@code Oracle} paging
+	 * {@link SafeOraclePagingQueryProvider provider} with an implementation that properly handles sort aliases.
+	 * <p><b>NOTE:</b> nested within the aggregate DAO as this is the only place that needs this specialization.
+	 */
+	static class SafeSqlPagingQueryProviderFactoryBean extends SqlPagingQueryProviderFactoryBean {
+
+		private DataSource dataSource;
+
+		@Override
+		public void setDataSource(DataSource dataSource) {
+			super.setDataSource(dataSource);
+			this.dataSource = dataSource;
+		}
+
+		@Override
+		public PagingQueryProvider getObject() throws Exception {
+			PagingQueryProvider provider = super.getObject();
+			if (provider instanceof OraclePagingQueryProvider) {
+				provider = new SafeOraclePagingQueryProvider((AbstractSqlPagingQueryProvider) provider, this.dataSource);
+			}
+			else if (provider instanceof SqlServerPagingQueryProvider) {
+				provider = new SafeSqlServerPagingQueryProvider((SqlServerPagingQueryProvider) provider, this.dataSource);
+			}
+			return provider;
+		}
+
+	}
+
+	/**
+	 * A {@link AbstractSqlPagingQueryProvider paging provider} for {@code Oracle} that works around the fact that the
+	 * Oracle provider in Spring Batch 4.x does not properly handle sort aliases when using nested {@code ROW_NUMBER}
+	 * clauses.
+	 */
+	static class SafeOraclePagingQueryProvider extends AbstractSqlPagingQueryProvider {
+
+		SafeOraclePagingQueryProvider(AbstractSqlPagingQueryProvider delegate, DataSource dataSource) {
+			// Have to use reflection to retrieve the provider fields
+			this.setFromClause(extractField(delegate, "fromClause", String.class));
+			this.setWhereClause(extractField(delegate, "whereClause", String.class));
+			this.setSortKeys(extractField(delegate, "sortKeys", Map.class));
+			this.setSelectClause(extractField(delegate, "selectClause", String.class));
+			this.setGroupClause(extractField(delegate, "groupClause", String.class));
+			try {
+				this.init(dataSource);
+			}
+			catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		private <T> T extractField(AbstractSqlPagingQueryProvider target, String fieldName, Class<T> fieldType) {
+			Field field = ReflectionUtils.findField(AbstractSqlPagingQueryProvider.class, fieldName, fieldType);
+			ReflectionUtils.makeAccessible(field);
+			return (T) ReflectionUtils.getField(field, target);
+		}
+
+		@Override
+		public String generateFirstPageQuery(int pageSize) {
+			return generateRowNumSqlQuery(false, pageSize);
+		}
+
+		@Override
+		public String generateRemainingPagesQuery(int pageSize) {
+			return generateRowNumSqlQuery(true, pageSize);
+		}
+
+		@Override
+		public String generateJumpToItemQuery(int itemIndex, int pageSize) {
+			int page = itemIndex / pageSize;
+			int offset = (page * pageSize);
+			offset = (offset == 0) ? 1 : offset;
+			String sortKeyInnerSelect = this.getSortKeySelect(true);
+			String sortKeyOuterSelect = this.getSortKeySelect(false);
+			return SqlPagingQueryUtils.generateRowNumSqlQueryWithNesting(this, sortKeyInnerSelect, sortKeyOuterSelect,
+					false, "TMP_ROW_NUM = " + offset);
+		}
+
+		private String getSortKeySelect(boolean withAliases) {
+			StringBuilder sql = new StringBuilder();
+			Map<String, Order> sortKeys = (withAliases) ? this.getSortKeys() : this.getSortKeysWithoutAliases();
+			sql.append(sortKeys.keySet().stream().collect(Collectors.joining(",")));
+			return sql.toString();
+		}
+
+		// Taken from SqlPagingQueryUtils.generateRowNumSqlQuery but use sortKeysWithoutAlias
+		// for outer sort condition.
+		private String generateRowNumSqlQuery(boolean remainingPageQuery, int pageSize) {
+			StringBuilder sql = new StringBuilder();
+			sql.append("SELECT * FROM (SELECT ").append(getSelectClause());
+			sql.append(" FROM ").append(this.getFromClause());
+			if (StringUtils.hasText(this.getWhereClause())) {
+				sql.append(" WHERE ").append(this.getWhereClause());
+			}
+			if (StringUtils.hasText(this.getGroupClause())) {
+				sql.append(" GROUP BY ").append(this.getGroupClause());
+			}
+			// inner sort by
+			sql.append(" ORDER BY ").append(SqlPagingQueryUtils.buildSortClause(this));
+			sql.append(") WHERE ").append("ROWNUM <= " + pageSize);
+			if (remainingPageQuery) {
+				sql.append(" AND ");
+				// For the outer sort we want to use sort keys w/o aliases. However,
+				// SqlPagingQueryUtils.buildSortConditions does not allow sort keys to be passed in.
+				// Therefore, we temporarily set the 'sortKeys' for the call to 'buildSortConditions'.
+				// The alternative is to clone the 'buildSortConditions' method here and allow the sort keys to be
+				// passed in BUT method is gigantic and this approach is the lesser of the two evils.
+				Map<String, Order> originalSortKeys = this.getSortKeys();
+				this.setSortKeys(this.getSortKeysWithoutAliases());
+				try {
+					SqlPagingQueryUtils.buildSortConditions(this, sql);
+				}
+				finally {
+					this.setSortKeys(originalSortKeys);
+				}
+			}
+			return sql.toString();
+		}
+	}
+
+	/**
+	 * A {@link SqlServerPagingQueryProvider paging provider} for {@code MSSQL} that works around the fact that the
+	 * MSSQL provider in Spring Batch 4.x does not properly handle sort aliases when generating jump to page queries.
+	 */
+	static class SafeSqlServerPagingQueryProvider extends SqlServerPagingQueryProvider {
+
+		SafeSqlServerPagingQueryProvider(SqlServerPagingQueryProvider delegate, DataSource dataSource) {
+			// Have to use reflection to retrieve the provider fields
+			this.setFromClause(extractField(delegate, "fromClause", String.class));
+			this.setWhereClause(extractField(delegate, "whereClause", String.class));
+			this.setSortKeys(extractField(delegate, "sortKeys", Map.class));
+			this.setSelectClause(extractField(delegate, "selectClause", String.class));
+			this.setGroupClause(extractField(delegate, "groupClause", String.class));
+			try {
+				this.init(dataSource);
+			}
+			catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		private <T> T extractField(AbstractSqlPagingQueryProvider target, String fieldName, Class<T> fieldType) {
+			Field field = ReflectionUtils.findField(AbstractSqlPagingQueryProvider.class, fieldName, fieldType);
+			ReflectionUtils.makeAccessible(field);
+			return (T) ReflectionUtils.getField(field, target);
+		}
+
+		@Override
+		protected String getOverClause() {
+			// Overrides the parent impl to use 'getSortKeys' instead of 'getSortKeysWithoutAliases'
+			StringBuilder sql = new StringBuilder();
+			sql.append(" ORDER BY ").append(SqlPagingQueryUtils.buildSortClause(this.getSortKeys()));
+			return sql.toString();
+		}
+
 	}
 }

--- a/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/support/SqlPagingQueryProviderFactoryBean.java
+++ b/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/support/SqlPagingQueryProviderFactoryBean.java
@@ -52,7 +52,7 @@ public class SqlPagingQueryProviderFactoryBean implements FactoryBean<PagingQuer
 	private final static Map<DatabaseType, AbstractSqlPagingQueryProvider> providers;
 
 	static {
-		 Map<DatabaseType, AbstractSqlPagingQueryProvider> providerMap = new HashMap<DatabaseType, AbstractSqlPagingQueryProvider>();
+		Map<DatabaseType, AbstractSqlPagingQueryProvider> providerMap = new HashMap<DatabaseType, AbstractSqlPagingQueryProvider>();
 		providerMap.put(DatabaseType.HSQL, new HsqlPagingQueryProvider());
 		providerMap.put(DatabaseType.H2, new H2PagingQueryProvider());
 		providerMap.put(DatabaseType.MYSQL, new MySqlPagingQueryProvider());

--- a/spring-cloud-dataflow-server/pom.xml
+++ b/spring-cloud-dataflow-server/pom.xml
@@ -408,5 +408,15 @@
 				</dependency>
 			</dependencies>
 		</profile>
+		<profile>
+			<id>local-dev-db2</id>
+			<dependencies>
+				<dependency>
+					<groupId>com.ibm.db2</groupId>
+					<artifactId>jcc</artifactId>
+					<version>11.5.8.0</version>
+				</dependency>
+			</dependencies>
+		</profile>
 	</profiles>
 </project>

--- a/spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/db/migration/AbstractSmokeTest.java
+++ b/spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/db/migration/AbstractSmokeTest.java
@@ -144,6 +144,8 @@ public abstract class AbstractSmokeTest {
 		createdExecutionIdsBySchemaTarget.add(schemaVersionTarget, execution1.getExecutionId());
 		TaskExecution execution2 = testUtils.createSampleJob("job2", 3, BatchStatus.COMPLETED, new JobParameters(), schemaVersionTarget);
 		createdExecutionIdsBySchemaTarget.add(schemaVersionTarget, execution2.getExecutionId());
+
+		// Get all executions and ensure the count and that the row number function was (or not) used
 		jobExecutions = taskJobService.listJobExecutionsWithStepCount(Pageable.ofSize(100));
 		assertThat(jobExecutions).hasSize(originalCount + 4);
 		String expectedSqlFragment = (this.supportsRowNumberFunction()) ?
@@ -151,6 +153,12 @@ public abstract class AbstractSmokeTest {
 				"as STEP_COUNT FROM AGGREGATE_JOB_INSTANCE";
 		Awaitility.waitAtMost(Duration.ofSeconds(5))
 				.untilAsserted(() -> assertThat(output).contains(expectedSqlFragment));
+
+		// Verify that paging works as well
+		jobExecutions = taskJobService.listJobExecutionsWithStepCount(Pageable.ofSize(2).withPage(0));
+		assertThat(jobExecutions).hasSize(2);
+		jobExecutions = taskJobService.listJobExecutionsWithStepCount(Pageable.ofSize(2).withPage(1));
+		assertThat(jobExecutions).hasSize(2);
 	}
 
 	static Stream<SchemaVersionTarget> schemaVersionTargetsProvider() {

--- a/spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/db/migration/JobExecutionTestUtils.java
+++ b/spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/server/db/migration/JobExecutionTestUtils.java
@@ -24,15 +24,22 @@ import java.util.Date;
 
 import javax.sql.DataSource;
 
+import com.zaxxer.hikari.HikariDataSource;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
 import org.springframework.batch.core.BatchStatus;
 import org.springframework.batch.core.JobExecution;
 import org.springframework.batch.core.JobInstance;
 import org.springframework.batch.core.JobParameters;
 import org.springframework.batch.core.repository.dao.JdbcJobInstanceDao;
 import org.springframework.batch.item.database.support.DataFieldMaxValueIncrementerFactory;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties;
 import org.springframework.cloud.dataflow.core.database.support.DatabaseType;
 import org.springframework.cloud.dataflow.core.database.support.MultiSchemaIncrementerFactory;
 import org.springframework.cloud.dataflow.schema.SchemaVersionTarget;
+import org.springframework.cloud.dataflow.schema.service.SchemaService;
+import org.springframework.cloud.dataflow.schema.service.impl.DefaultSchemaService;
 import org.springframework.cloud.dataflow.server.repository.TaskBatchDaoContainer;
 import org.springframework.cloud.dataflow.server.repository.TaskExecutionDaoContainer;
 import org.springframework.cloud.task.batch.listener.TaskBatchDao;
@@ -129,5 +136,38 @@ class JobExecutionTestUtils
 
 	private Timestamp timestampFromDate(Date date) {
 		return (date != null) ? Timestamp.valueOf(date.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()) : null;
+	}
+
+
+	/**
+	 * Test utility that generates hundreds of job executions which can be useful when debugging paging issues.
+	 * <p>To run, adjust the datasource properties accordingly and then execute the test manually in your editor.
+	 */
+	@Disabled
+	static class JobExecutionTestDataGenerator {
+
+		@Test
+		void generateJobExecutions() {
+			// Adjust these properties as necessary to point to your env
+			DataSourceProperties dataSourceProperties = new DataSourceProperties();
+			dataSourceProperties.setUrl("jdbc:oracle:thin:@localhost:1521/dataflow");
+			dataSourceProperties.setUsername("spring");
+			dataSourceProperties.setPassword("spring");
+			dataSourceProperties.setDriverClassName("oracle.jdbc.OracleDriver");
+
+			DataSource dataSource = dataSourceProperties.initializeDataSourceBuilder().type(HikariDataSource.class).build();
+			SchemaService schemaService = new DefaultSchemaService();
+			TaskExecutionDaoContainer taskExecutionDaoContainer = new TaskExecutionDaoContainer(dataSource, schemaService);
+			TaskBatchDaoContainer taskBatchDaoContainer = new TaskBatchDaoContainer(dataSource, schemaService);
+			JobExecutionTestUtils generator = new JobExecutionTestUtils(taskExecutionDaoContainer, taskBatchDaoContainer);
+			generator.createSampleJob(jobName("boot2"), 200, BatchStatus.COMPLETED, new JobParameters(),
+					schemaService.getTarget("boot2"));
+			generator.createSampleJob(jobName("boot3"), 200, BatchStatus.COMPLETED, new JobParameters(),
+					schemaService.getTarget("boot3"));
+		}
+
+		private String jobName(String schemaTarget) {
+			return schemaTarget + "-job-" + System.currentTimeMillis();
+		}
 	}
 }

--- a/spring-cloud-dataflow-test/src/main/java/org/springframework/cloud/dataflow/server/db/DB2_11_5_ContainerSupport.java
+++ b/spring-cloud-dataflow-test/src/main/java/org/springframework/cloud/dataflow/server/db/DB2_11_5_ContainerSupport.java
@@ -16,10 +16,12 @@
 
 package org.springframework.cloud.dataflow.server.db;
 
-import org.testcontainers.containers.Db2Container;
-import org.testcontainers.junit.jupiter.Container;
-import org.testcontainers.junit.jupiter.Testcontainers;
+import java.util.concurrent.atomic.AtomicReference;
 
+import org.junit.jupiter.api.BeforeAll;
+import org.testcontainers.containers.Db2Container;
+
+import org.springframework.cloud.dataflow.server.db.arm64.Db2Arm64ContainerSupport;
 import org.springframework.test.context.DynamicPropertyRegistry;
 import org.springframework.test.context.DynamicPropertySource;
 
@@ -28,14 +30,20 @@ import org.springframework.test.context.DynamicPropertySource;
  *
  * @author Chris Bono
  */
-@Testcontainers(disabledWithoutDocker = true)
-public interface DB2_11_5_ContainerSupport {
+public interface DB2_11_5_ContainerSupport extends Db2Arm64ContainerSupport {
 
-	@Container
-	Db2Container container = new Db2Container("ibmcom/db2:11.5.0.0a").acceptLicense();
+	AtomicReference<Db2Container> containerReference = new AtomicReference<>(null);
+
+	@BeforeAll
+	static void startContainer() {
+		Db2Container container = Db2Arm64ContainerSupport.startContainer(() ->
+				new Db2Container("ibmcom/db2:11.5.0.0a").acceptLicense());
+		containerReference.set(container);
+	}
 
 	@DynamicPropertySource
 	static void databaseProperties(DynamicPropertyRegistry registry) {
+		Db2Container container = containerReference.get();
 		registry.add("spring.datasource.url", container::getJdbcUrl);
 		registry.add("spring.datasource.username", container::getUsername);
 		registry.add("spring.datasource.password", container::getPassword);

--- a/spring-cloud-dataflow-test/src/main/java/org/springframework/cloud/dataflow/server/db/arm64/Db2Arm64ContainerSupport.java
+++ b/spring-cloud-dataflow-test/src/main/java/org/springframework/cloud/dataflow/server/db/arm64/Db2Arm64ContainerSupport.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2023-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.dataflow.server.db.arm64;
+
+import java.util.function.Supplier;
+
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.testcontainers.containers.Db2Container;
+import org.testcontainers.containers.OracleContainer;
+import uk.org.webcompere.systemstubs.environment.EnvironmentVariables;
+import uk.org.webcompere.systemstubs.jupiter.SystemStub;
+import uk.org.webcompere.systemstubs.jupiter.SystemStubsExtension;
+
+import org.springframework.cloud.dataflow.server.db.ContainerSupport;
+import org.springframework.core.log.LogAccessor;
+
+/**
+ * Provides support for testing against an {@link Db2Container DB2 testcontainer} on Mac ARM64.
+ *
+ * @author Chris Bono
+ */
+@ExtendWith(SystemStubsExtension.class)
+public interface Db2Arm64ContainerSupport {
+
+	LogAccessor LOG = new LogAccessor(Db2Arm64ContainerSupport.class);
+
+	@SystemStub
+	EnvironmentVariables ENV_VARS = new EnvironmentVariables();
+
+	static Db2Container startContainer(Supplier<Db2Container> db2ContainerSupplier) {
+		if (ContainerSupport.runningOnMacArm64()) {
+			String wiki = "https://github.com/spring-cloud/spring-cloud-dataflow/wiki/DB2-on-Mac-ARM64";
+			LOG.warn(() -> "You are running on Mac ARM64. If this test fails, make sure Colima is running prior " +
+					"to test invocation. See " + wiki + " for details");
+			ENV_VARS.set("TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE", "/var/run/docker.sock");
+			ENV_VARS.set("DOCKER_HOST", String.format("unix://%s/.colima/docker.sock", System.getProperty("user.home")));
+		}
+		Db2Container container = db2ContainerSupplier.get();
+		container.start();
+		return container;
+	}
+}

